sdk/src/azure/platform/az_pahoasync.c

```diff
@@ -0,0 +1,619 @@
+
+  az_result ret = az_mqtt5_inbound_disconnect(
+      me,
+      &(az_mqtt5_disconnect_data){ .tls_authentication_error = false,
+                                   .disconnect_requested = (response->reasonCode == 0) });
+
+  if (az_result_failed(ret))
+  {
+    _az_pahoasync_critical_error();
+  }
+}
+
+static void _az_pahoasync_on_disconnect_failure(void* context, MQTTAsync_failureData5* response)
+{
+  (void)context;
+  (void)response;
+
+  // TODO_L: Implement logic to handle failing to disconnect.
+  _az_pahoasync_critical_error();
+}
+
+AZ_NODISCARD az_mqtt5_options az_mqtt5_options_default()
+{
+  return (az_mqtt5_options){
+    .certificate_authority_trusted_roots = AZ_SPAN_EMPTY,
+    .disable_tls = false,
+  };
+}
+
+AZ_NODISCARD az_result
+az_mqtt5_init(az_mqtt5* mqtt5, MQTTAsync* pahoasync_handle, az_mqtt5_options const* options)
+{
+  _az_PRECONDITION_NOT_NULL(mqtt5);
+  _az_PRECONDITION_NOT_NULL(pahoasync_handle);
+
+  mqtt5->_internal.options = options == NULL ? az_mqtt5_options_default() : *options;
+  mqtt5->_internal.pahoasync_handle = pahoasync_handle;
+  mqtt5->_internal.platform_mqtt5.pipeline = NULL;
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result
+az_mqtt5_outbound_connect(az_mqtt5* mqtt5, az_mqtt5_connect_data* connect_data)
+{
+  _az_PRECONDITION_NOT_NULL(mqtt5);
+  _az_PRECONDITION_NOT_NULL(connect_data);
+  _az_PRECONDITION_VALID_SPAN(connect_data->host, 1, false);
+  _az_PRECONDITION_VALID_SPAN(connect_data->client_id, 1, false);
+
+  az_result ret = AZ_OK;
+  az_mqtt5* me = (az_mqtt5*)mqtt5;
+  MQTTAsync_createOptions create_opts = MQTTAsync_createOptions_initializer5;
+  MQTTAsync_connectOptions conn_opts = MQTTAsync_connectOptions_initializer5;
+  MQTTAsync_SSLOptions ssl_opts = MQTTAsync_SSLOptions_initializer;
+
+  if (me->_internal.options.disable_tls == 0)
+  {
+    ssl_opts.verify = 1;
+    bool use_os_certs
+        = (az_span_ptr(me->_internal.options.certificate_authority_trusted_roots) == NULL);
+
+    if (!use_os_certs)
+    {
+      ssl_opts.disableDefaultTrustStore = 1;
+      ssl_opts.CApath
+          = (const char*)az_span_ptr(me->_internal.options.certificate_authority_trusted_roots);
+    }
+
+    ssl_opts.keyStore = (const char*)az_span_ptr(connect_data->certificate.cert);
+    ssl_opts.privateKey = (const char*)az_span_ptr(connect_data->certificate.key);
+  }
+
+  if (az_span_ptr(connect_data->username) != NULL)
+  {
+    conn_opts.username = (const char*)az_span_ptr(connect_data->username);
+    conn_opts.password = (const char*)az_span_ptr(connect_data->password);
+  }
+
+  char port_buffer[SERVER_URI_PORT_MAX_LENGTH];
+  sprintf(port_buffer, "%d", connect_data->port);
+  az_span port_span = az_span_create_from_str(port_buffer);
+
+  int32_t server_uri_length
+      = az_span_size(connect_data->host) + SERVER_URI_COLON_LENGTH + SERVER_URI_PORT_MAX_LENGTH + 1;
+  char server_uri_buffer[server_uri_length];
+  az_span server_uri = az_span_create((uint8_t*)server_uri_buffer, server_uri_length);
```

Comment:
```suggestion
  az_span server_uri = az_span_create_from_buffer((uint8_t*)server_uri_buffer);
```

Reply:
There is no `az_span_create_from_buffer` but if you're referring to `AZ_SPAN_FROM_BUFFER`, I already tried that approach but obtain a compiler error on Ubuntu 20 because of the conversion I do for the length of the buffer (from unsigned int to int)